agent/internal/taskoutput/test_log.go

```diff
@@ -0,0 +1,330 @@
+	for {
+		err := filepath.WalkDir(h.dir, func(path string, info fs.DirEntry, err error) error {
+			if err != nil {
+				h.logger.Execution().Warning(errors.Wrap(err, "watching the test log directory, continuing watching"))
+				return nil
+			}
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			if info.IsDir() {
+				return nil
+			}
+
+			if !seenFiles[path] {
+				seenFiles[path] = true
+				h.followFileCount++
+				h.wg.Add(1)
+				go h.followFile(workerCtx, path)
+			}
+
+			return nil
+		})
+		if err != nil {
+			h.logger.Execution().Debug("context canceled, exiting test log directory watcher")
+			return
+		}
+	}
+}
+
+// getSpecFile looks for the test log specification file in the top level of
+// the reserved test log directory. If the spec file cannot be read for any
+// reason, an error is logged and the handler uses the default spec.
+//
+// Called once per task run immediately after the first log file is detected
+// and before any data is ingested.
+func (h *testLogDirectoryHandler) getSpecFile() {
+	h.logger.Task().Infof("detected first test log file, getting test log spec file")
+
+	data, err := os.ReadFile(filepath.Join(h.dir, testLogSpecFilename))
+	if err != nil {
+		h.logger.Task().Warning(errors.Wrap(err, "reading test log spec; falling back to default spec"))
+		return
+	}
+	if err = yaml.Unmarshal(data, &h.spec); err != nil {
+		h.logger.Task().Warning(errors.Wrap(err, "unmarshalling test log spec; falling back to default spec"))
+		return
+	}
+
+	if err = h.spec.Format.validate(); err != nil {
+		h.logger.Task().Warning(errors.Wrapf(err, "invalid test log format specified; falling back to default text format"))
+	}
+}
+
+// followFile tails a test log file for the duration of a task, ingesting and
+// persisting log lines as they are written.
+func (h *testLogDirectoryHandler) followFile(ctx context.Context, path string) {
+	defer func() {
+		h.logger.Task().Critical(recovery.HandlePanicWithError(recover(), nil, "test log file follower"))
+	}()
+	defer h.wg.Done()
+
+	h.once.Do(h.getSpecFile)
+
+	h.logger.Task().Infof("new test log file '%s' found, initiating automated ingestion", path)
+
+	// The persisted log path should be relative to the reserved directory.
+	logPath, err := filepath.Rel(h.dir, path)
+	if err != nil {
+		h.logger.Task().Error(errors.Wrapf(err, "getting relative path for test log file '%s'", path))
+		return
+	}
+
+	t, err := tail.TailFile(path, tail.Config{
+		ReOpen:    true,
+		MustExist: true,
+		Follow:    true,
+		// Setting this field may lead to some data loss in the case of
+		// non-atomic line writes: if the last read line does not
+		// terminate with a newline character (e.g., the follower
+		// exits early or the test process crashes), it will not get
+		// returned. This is ok because the complete line must be read
+		// to guarantee proper line format for the parser (assumming
+		// the complete line has the expected format).
+		CompleteLines: true,
```

Comment:
does this mean the last line in the file will only be read if the file is terminated with a \n?
Like in a regular case: I wrote a file and exited before the task finishes.

Reply:
Exactly, all log lines written in the "happy path" scenario should be terminated with \n.

Reply:
But are we comfortable with not potentially not getting the last line?
Could we want `close` to read the rest of the file from the position of the last read?

Reply:
Yeah, log lines should always end in a newline so we would only not get it in the case that something went wrong. If we don't get complete lines, we risk a parsing a malformed line and breaking the parser, in which case, the line would fail to get persisted anyways.

Reply:
This will also be documented for users so they understand the behavior.